runtime/compiler/runtime/IProfiler.cpp

```diff
@@ -4853,67 +4901,309 @@ void TR_IProfiler::dumpIPBCDataCallGraph(J9VMThread* vmThread)
+   {
+   if (high < low)
+      return 0;
 
-   fprintf(stderr, "Aggregating per method ...\n");
-   for (int32_t bucket = 0; bucket < TR::Options::_iProfilerBcHashTableSize; bucket++)
+   TR_IPBCDataStorageHeader * storage = (TR_IPBCDataStorageHeader *) memChunk;
+   int32_t middle = (high+low)/2;
+   TR_IPBytecodeHashTableEntry *entry = ipEntries[middle];
+   uint32_t bytes = entry->getBytesFootprint();
+   entry->createPersistentCopy(sharedCache, storage, _compInfo->getPersistentInfo());
+
+   uintptr_t leftChild = createBalancedBST(ipEntries, low, middle - 1, memChunk + bytes, sharedCache);
+   if (leftChild)
       {
-      //fprintf(stderr, "Looking at bucket %d\n", bucket);
-      for (TR_IPBytecodeHashTableEntry *entry = _bcHashTable[bucket]; entry; entry = entry->getNext())
+      TR_ASSERT(bytes < 1 << 8, "Error storing iprofile information: left child too far away"); // current size of left child
+      storage->left = bytes;
+      }
+
+   uintptr_t rightChild = createBalancedBST(ipEntries, middle + 1, high, memChunk + bytes + leftChild, sharedCache);
+   if (rightChild)
+      {
+      TR_ASSERT(bytes + leftChild < 1 << 16, "Error storing iprofile information: right child too far away"); // current size of right child
+      storage->right = bytes+leftChild;
+      }
+
+   return bytes + leftChild + rightChild;
+   }
+
+// Persist all IProfiler entries into the SCC.
+// This is done by aggregating all IProfiler entries per
+// ROMMethod with the help of a TR_AggregationHT table.
+// Then, for each ROMMethod with IP info we store the
+// entries arranged as a BST (for fast retrieval later).
+void
+TR_IProfiler::persistAllEntries()
+   {
+   J9JavaVM * javaVM = _compInfo->getJITConfig()->javaVM;
+   J9VMThread *vmThread = javaVM->internalVMFunctions->currentVMThread(javaVM);
+   TR_J9VMBase * fe = TR_J9VMBase::get(_compInfo->getJITConfig(), vmThread, TR_J9VMBase::AOT_VM);
+   TR_J9SharedCache *sharedCache = fe->sharedCache();
+   static bool SCfull = false;
+
+   if (!(TR::Options::sharedClassCache() && sharedCache))
+      return;
+
+   //int locked = releaseAllEntries();
+   J9SharedClassConfig *scConfig = javaVM->sharedClassConfig;
+
+   if (TR::Options::getCmdLineOptions()->getVerboseOption(TR_VerboseIProfilerPersistence))
+      TR_VerboseLog::writeLineLocked(TR_Vlog_PERF, "IProfiler persisting all entries to SCC");
+
+   int32_t entriesAlreadyPersisted = _STATS_entriesPersisted;
+
+   try
+      {
+      TR::RawAllocator rawAllocator(javaVM);
+      J9::SegmentAllocator segmentAllocator(MEMORY_TYPE_JIT_SCRATCH_SPACE | MEMORY_TYPE_VIRTUAL, *javaVM);
+      J9::SystemSegmentProvider regionSegmentProvider(1 << 20, 1 << 20, TR::Options::getScratchSpaceLimit(), segmentAllocator, rawAllocator);
+      TR::Region region(regionSegmentProvider, rawAllocator);
+      TR_Memory trMemory(*_compInfo->persistentMemory(), region);
+
+      // TODO: use heap memory
```

Comment:
What heap memory is this referring to?

Reply:
It refers to compiler heap memory as opposed to persistent memory. Persistent memory is fine if you use this function once, during shutdown. If you want to save IProfiler more often, or in the middle of the run, then memory that can be released without causing fragmentation is better.